src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,919 @@
+(define-constant ERR_NAMESPACE_PREORDER_CLAIMABILITY_EXPIRED 1009)
+(define-constant ERR_NAMESPACE_PREORDER_LAUNCHABILITY_EXPIRED 1010)
+(define-constant ERR_NAMESPACE_OPERATION_UNAUTHORIZED 1011)
+(define-constant ERR_NAMESPACE_STX_BURNT_INSUFFICIENT 1012)
+(define-constant ERR_NAMESPACE_BLANK 1013)
+(define-constant ERR_NAMESPACE_ALREADY_LAUNCHED 1014)
+(define-constant ERR_NAMESPACE_HASH_MALFORMED 1015)
+(define-constant ERR_NAMESPACE_CHARSET_INVALID 1016)
+
+(define-constant ERR_NAME_PREORDER_NOT_FOUND 2001)
+(define-constant ERR_NAME_PREORDER_EXPIRED 2002)
+(define-constant ERR_NAME_PREORDER_FUNDS_INSUFFICIENT 2003)
+(define-constant ERR_NAME_UNAVAILABLE 2004)
+(define-constant ERR_NAME_OPERATION_UNAUTHORIZED 2006)
+(define-constant ERR_NAME_STX_BURNT_INSUFFICIENT 2007)
+(define-constant ERR_NAME_EXPIRED 2008)
+(define-constant ERR_NAME_GRACE_PERIOD 2009)
+(define-constant ERR_NAME_BLANK 2010)
+(define-constant ERR_NAME_ALREADY_CLAIMED 2011)
+(define-constant ERR_NAME_CLAIMABILITY_EXPIRED 2012)
+(define-constant ERR_NAME_NOT_FOUND 2013)
+(define-constant ERR_NAME_REVOKED 2014)
+(define-constant ERR_NAME_TRANSFER_FAILED 2015)
+(define-constant ERR_NAME_PREORDER_ALREADY_EXISTS 2016)
+(define-constant ERR_NAME_HASH_MALFORMED 2017)
+(define-constant ERR_NAME_PREORDERED_BEFORE_NAMESPACE_LAUNCH 2018)
+(define-constant ERR_NAME_NOT_RESOLVABLE 2019)
+(define-constant ERR_NAME_COULD_NOT_BE_MINTED 2020)
+(define-constant ERR_NAME_COULD_NOT_BE_TRANSFERED 2021)
+(define-constant ERR_NAME_CHARSET_INVALID 2022)
+
+(define-constant ERR_PRINCIPAL_ALREADY_ASSOCIATED 3001)
+(define-constant ERR_INSUFFICIENT_FUNDS 4001)
+
+(define-constant NAMESPACE_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAMESPACE_LAUNCHABILITY_TTL u10)
+(define-constant NAME_PREORDER_CLAIMABILITY_TTL u10)
+(define-constant NAME_GRACE_PERIOD_DURATION u5)
+
+(define-constant ATTACHMENTS_INV_PAGE_SIZE u8)
+(define-data-var attachments-inv-index-cursor uint u0)
+(define-data-var attachments-inv-page-cursor uint u0)
+(define-map attachments-inv 
+    ((page uint) (index uint)) 
+    ((content-hash (buff 20))))
+
+;; Price tables
+(define-constant NAMESPACE_PRICE_TIERS (list
+  u96000 
+  u9600 u9600 
+  u960 u960 u960 u960 
+  u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96 u96))
+
+;;;; Data
+(define-map namespaces
+  ((namespace (buff 20)))
+  ((namespace-import principal)
+   (revealed-at uint)
+   (launched-at (optional uint))
+   (lifetime uint)
+   (price-function (tuple 
+    (buckets (list 16 uint)) 
+    (base uint) 
+    (coeff uint) 
+    (nonalpha-discount uint) 
+    (no-vowel-discount uint)))))
+
+(define-map namespace-preorders
+  ((hashed-salted-namespace (buff 20)) (buyer principal))
+  ((created-at uint) (claimed bool) (stx-burned uint)))
+
+(define-non-fungible-token names (tuple (name (buff 32)) (namespace (buff 20))))
+
+;; Rule 1-1 -> 1 principal, 1 name
+(define-map rule-1-1 ((owner principal)) ((name (buff 32)) (namespace (buff 20))))
```

Comment:
Not a bug, but why not call this map `owner-name`, or something that conveys what the keys and values are?  I had to double-check what `rule-1-1` meant when I encountered it later.  Normally I wouldn't be picky on this, but once this code is deployed, we can't change it.

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/f80dc6564552038e47487654f9dce0c760d81033